cluster-agent/controllers/argoproj.io/namespace_reconciler.go

```diff
@@ -0,0 +1,335 @@
+		log.Info("Prune field in ArgoCD and DB entry is not in Sync.")
+		return false
+	}
+	if applicationFromArgoCD.Spec.SyncPolicy.Automated.SelfHeal != applicationFromDB.Spec.SyncPolicy.Automated.SelfHeal {
+		log.Info("SelfHeal field in ArgoCD and DB entry is not in Sync.")
+		return false
+	}
+	if applicationFromArgoCD.Spec.SyncPolicy.Automated.AllowEmpty != applicationFromDB.Spec.SyncPolicy.Automated.AllowEmpty {
+		log.Info("AllowEmpty field in ArgoCD and DB entry is not in Sync.")
+		return false
+	}
+	return true
+}
+
+func cleanK8sOperations(ctx context.Context, dbq db.DatabaseQueries, client client.Client, log logr.Logger) {
+	// Get list of Operations from cluster.
+	listOfK8sOperation := v1alpha1.OperationList{}
+	err := client.List(ctx, &listOfK8sOperation)
+	if err != nil {
+		log.Error(err, "Unable to fetch list of k8s Operation from cluster.")
+		return
+	}
+
+	for _, k8sOperation := range listOfK8sOperation.Items {
+
+		// Skip if Operation was not created by Namespace Reconciler.
+		if k8sOperation.Annotations[appEventLoop.IdentifierKey] != appEventLoop.IdentifierValue {
+			continue
+		}
+
+		// Fetch corresponding DB entry
+		dbOperation := db.Operation{
+			Operation_id: k8sOperation.Spec.OperationID,
+		}
+		if err := dbq.GetOperationById(ctx, &dbOperation); err != nil {
+			continue
+		}
+
+		if dbOperation.State != db.OperationState_Completed && dbOperation.State != db.OperationState_Failed {
+			log.V(sharedutil.LogLevel_Debug).Info("K8s Operation is not ready for cleanup : " + string(k8sOperation.UID) + " DbOperation: " + string(k8sOperation.Spec.OperationID))
+			continue
+		}
+
+		log.Info("Deleting Operation created by Namespace Reconciler." + string(k8sOperation.UID))
+
+		// Delete the k8s operation now.
+		if err := appEventLoop.CleanupOperation(ctx, dbOperation, k8sOperation, cache.GetGitOpsEngineSingleInstanceNamespace(), dbq, client, log); err != nil {
+			log.Error(err, "Unable to Delete k8s Operation"+string(k8sOperation.UID)+" for DbOperation: "+string(k8sOperation.Spec.OperationID))
+		} else {
+			log.Info("Deleted k8s Operation: " + string(k8sOperation.UID) + " for DbOperation: " + string(k8sOperation.Spec.OperationID))
+		}
+	}
+	log.V(sharedutil.LogLevel_Debug).Info("Cleaned all Operations created by Namespace Reconciler.")
+}
+
+func deleteOrphanedApplications(argoApplications []appv1.Application, processedApplicationIds map[string]string, ctx context.Context, client client.Client, log logr.Logger) []appv1.Application {
+	log.V(sharedutil.LogLevel_Debug).Info("Looking for any orphaned ArgoCD application to delete.")
+
+	// Iterate through all application IDs collected from DB.
+	for id := range processedApplicationIds {
+		// Iterate through ArgoCD all and remove corresponding application from list.
+		for i, application := range argoApplications {
+			if application.Labels["databaseID"] == id {
+				argoApplications = dropProcessedApplicationId(argoApplications, i)
+			}
+		}
+	}
```

Comment:
Since we are using a map for `processedApplicationIds`, can we avoid iterating both the loops?

For example:

```
for _, application := range argoApplications {
    if _, ok := processedApplicationIds[application.Labels["databaseID"]]; !ok {
	    if err := controllers.DeleteArgoCDApplication(ctx, application, client, log); err != nil {
		    log.Error(err, "unable to delete an orphaned ArgoCD application "+application.Name)
	    } else {
		    log.Info("Deleted orphaned ArgoCD application " + application.Name)
	    }
    }
}
```
Also the map `processedApplicationIds` might be empty and we would still remove the orphaned applications.

Reply:
Instead of removing elements from the array and then deleting the remaining ones from DB inside `deleteApplication()`, what if we directly delete them from the DB inside the loop? Then we don't have to update the array right? The above example tries to do something similar

Reply:
My bad, I misunderstood here. Changed the logic as you suggested. PTAL.